programs/drift_vaults/src/state/vault_depositor.rs

```diff
@@ -12,13 +58,353 @@ pub struct VaultDepositor {
+        )?;
+
+        self.last_withdraw_request_shares = n_shares;
+        vault.apply_rebase(vault_amount)?;
+        self.apply_rebase(vault)?;
+
+        let vault_shares_before: u128 = self.checked_vault_shares(vault)?;
+        let total_vault_shares_before = vault.total_shares;
+        let user_vault_shares_before = vault.user_shares;
+
+        validate!(
+            self.last_withdraw_request_shares <= self.checked_vault_shares(vault)?,
+            ErrorCode::InvalidVaultWithdrawSize,
+            "last_withdraw_request_shares exceeds vault_shares {} > {}",
+            self.last_withdraw_request_shares,
+            self.checked_vault_shares(vault)?
+        )?;
+
+        validate!(
+            self.vault_shares_base == vault.shares_base,
+            ErrorCode::InvalidVaultRebase,
+            "vault depositor shares_base != vault shares_base"
+        )?;
+
+        self.last_withdraw_request_value = if_shares_to_vault_amount(
+            self.last_withdraw_request_shares,
+            vault.total_shares,
+            vault_amount,
+        )?
+        .min(vault_amount.saturating_sub(1));
+
+        validate!(
+            self.last_withdraw_request_value == 0
+                || self.last_withdraw_request_value < vault_amount,
+            ErrorCode::InvalidVaultWithdrawSize,
+            "Requested withdraw value is not below Insurance Fund balance"
+        )?;
+
+        let vault_shares_after = self.checked_vault_shares(vault)?;
+
+        emit!(VaultDepositorRecord {
+            ts: now,
+            vault: vault.pubkey,
+            user_authority: self.authority,
+            action: VaultDepositorAction::WithdrawRequest,
+            amount: self.last_withdraw_request_value,
+            spot_market_index: vault.spot_market_index,
+            vault_amount_before: vault_amount,
+            vault_shares_before,
+            user_vault_shares_before,
+            total_vault_shares_before,
+            vault_shares_after,
+            total_vault_shares_after: vault.total_shares,
+            user_vault_shares_after: vault.user_shares,
+        });
+
+        self.last_withdraw_request_ts = now;
+
+        Ok(self.last_withdraw_request_value)
+    }
+
+    pub fn cancel_withdraw_request(
+        self: &mut VaultDepositor,
+        vault_amount: u64,
+        vault: &mut Vault,
+        now: i64,
+    ) -> Result<()> {
+        vault.apply_rebase(vault_amount)?;
+        self.apply_rebase(vault)?;
+
+        let vault_shares_before: u128 = self.checked_vault_shares(vault)?;
+        let total_vault_shares_before = vault.total_shares;
+        let user_vault_shares_before = vault.user_shares;
+
+        validate!(
+            self.vault_shares_base == vault.shares_base,
+            ErrorCode::InvalidVaultRebase,
+            "vault depositor shares_base != vault shares_base"
+        )?;
+
+        let vault_shares_lost = calculate_vault_shares_lost(self, vault, vault_amount)?;
```

Comment:
might be nice to make this configurable for vault auth?

also should probably get a doc going w all the todos

Reply:
https://github.com/orgs/drift-labs/projects/2/views/1?pane=issue&itemId=29886547